src/ObjectReadStateMachine.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2021 Atif Aziz.
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Jacob;
+
+using System;
+using System.Text.Json;
+
+public record struct ObjectReadStateMachine
+{
+    public enum State { Initial, PropertyNameOrEnd, PendingPropertyNameRead, PendingPropertyValueRead, Done, Error }
+    public enum ReadResult { Error, Incomplete, PropertyName, PropertyValue, Done }
+
+    public State CurrentState { get; private set; }
+
+    public void OnPropertyNameRead() =>
+        CurrentState =
+            CurrentState is State.PendingPropertyNameRead
+            ? State.PendingPropertyValueRead
+            : throw new InvalidOperationException();
+
+    public void OnPropertyValueRead() =>
+        CurrentState =
+            CurrentState is State.PendingPropertyValueRead
+            ? State.PropertyNameOrEnd
+            : throw new InvalidOperationException();
+
+    public ReadResult Read(ref Utf8JsonReader reader)
+    {
+        while (true)
+        {
+            switch (CurrentState)
+            {
+                case State.Initial:
```

Comment:
Some `case` arms are multi-statement so I would put them all consistently in a block (`{`…`}`).

Reply:
Done in 7b8fc262061d129ba4df47354d9efb941ae61e22